### PR TITLE
Deploy Polygon Amoy, drop Mumbai

### PR DESCRIPTION
**Pull request by dtbuchholz**
## Summary

Adds a new utility function to get live gas prices for Amoy and deploys new registry contracts there. Also, drops Mumbai.

## Details

For some reason, ethers' built-in method `getFeeData` (which makes a call to get gas prices from a node) returns absurdly wrong data for Amoy. E.g., `maxFeePerGas` of `1000000030n` and `maxPriorityFeePerGas` of `1000000000n`—but in reality, these are 30x larger at `35185714305n` and `35185714290n`, respectively. Thus, an API call to the [gas station](https://gasstation-testnet.polygon.technology/amoy) is required. Note: there is a [built-in custom gas fee class](https://github.com/ethers-io/ethers.js/blob/29beb93ea76d560a129e4bc3fde6ee9c8cc887c4/src.ts/providers/plugins-network.ts#L200) in ethers, but I couldn't figure out how to use it.

## How it was tested

Deploying on live net.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: hardhat.config.ts [hunk 1]
```diff
@@ -68,7 +68,7 @@ const config: HardhatUserConfig = {
 
       // polygon
       polygon: process.env.POLYSCAN_API_KEY || "",
-      polygonMumbai: process.env.POLYSCAN_API_KEY || "",
+      polygonAmoy: process.env.OKLINK_API_KEY || "",
```
Comment (dtbuchholz) [inline hardhat.config.ts:71]: note: the contract verification didn't work—i can dig into it later but not required atm

File: scripts/utils.ts [hunk 2]
```diff
@@ -1,5 +1,61 @@
...
+ */
+export async function getFeeData(chainName: string): Promise<FeeData> {
+  if (chainName === "polygon-amoy") {
+    try {
+      const url = "https://gasstation-testnet.polygon.technology/amoy";
+      const response = await fetch(url);
+      const data = (await response.json()) as AmoyFeeData;
+      const feeData = new FeeData(
+        null, // No gas price value needed
+        BigInt(Math.ceil(data.fast.maxFee * 1e9)),
```
Comment (dtbuchholz) [inline scripts/utils.ts:40]: i _think_ this is the correct conversion from gwei to wei. e.g., the gas station will give you a value like `"maxPriorityFee": 301.778399988,
"maxFee": 301.778400003`

File: network.ts [hunk 3]
```diff
@@ -29,13 +29,13 @@ export const proxies: TablelandNetworkConfig = {
...
   "arbitrum-nova": "0x1A22854c5b1642760a827f20137a67930AE108d2",
-  matic: "0x5c4e6A9e5C1e1BF445A062006faF19EA6c49aFeA",
+  polygon: "0x5c4e6A9e5C1e1BF445A062006faF19EA6c49aFeA",
   filecoin: "0x59EF8Bf2d6c102B4c42AEf9189e1a9F0ABfD652d",
   // testnets
   sepolia: "0xc50C62498448ACc8dBdE43DA77f8D5D2E2c7597D",
   "optimism-sepolia": "0x68A2f4423ad3bf5139Db563CF3bC80aA09ed7079",
   "arbitrum-sepolia": "0x223A74B8323914afDC3ff1e5005564dC17231d6e",
-  maticmum: "0x4b48841d4b32C4650E4ABc117A03FE8B51f38F68",
+  "polygon-amoy": "0x170fb206132b693e38adFc8727dCfa303546Cec1",
```
Comment (dtbuchholz) [inline network.ts:38]: cc @avichalp

File: README.md [hunk 4]
```diff
@@ -26,12 +26,12 @@ This is the Tableland Tables EVM contracts and client components.
...
 | arbitrum             | 42161    | 0x9aBd75E8640871A5a20d3B4eE6330a04c962aFfd |
 | arbitrum-nova        | 42170    | 0x1A22854c5b1642760a827f20137a67930AE108d2 |
-| matic                | 137      | 0x5c4e6A9e5C1e1BF445A062006faF19EA6c49aFeA |
+| polygon              | 137      | 0x5c4e6A9e5C1e1BF445A062006faF19EA6c49aFeA |
 | filecoin             | 314      | 0x59EF8Bf2d6c102B4c42AEf9189e1a9F0ABfD652d |
 | sepolia              | 11155111 | 0xc50C62498448ACc8dBdE43DA77f8D5D2E2c7597D |
 | optimism-sepolia     | 11155420 | 0x68A2f4423ad3bf5139Db563CF3bC80aA09ed7079 |
 | arbitrum-sepolia     | 421614   | 0x223A74B8323914afDC3ff1e5005564dC17231d6e |
-| maticmum             | 80001    | 0x4b48841d4b32C4650E4ABc117A03FE8B51f38F68 |
+| polygon-amoy         | 80002    | 0x4b48841d4b32C4650E4ABc117A03FE8B51f38F68 |
```
Comment (sanderpick) [inline README.md:34]: need new address here

File: contracts/utils/TablelandDeployments.sol [hunk 5]
```diff
@@ -45,8 +45,8 @@ library TablelandDeployments {
     // TablelandTables address on Arbitrum Sepolia.
     address internal constant ARBITRUM_SEPOLIA =
         0x223A74B8323914afDC3ff1e5005564dC17231d6e;
-    // TablelandTables address on Polygon Mumbai.
-    address internal constant MATICMUM =
+    // TablelandTables address on Polygon Amoy.
+    address internal constant POLYGON_AMOY =
         0x4b48841d4b32C4650E4ABc117A03FE8B51f38F68;
```
Comment (sanderpick) [inline contracts/utils/TablelandDeployments.sol:50]: and here

File: hardhat.config.ts [hunk 6]
```diff
@@ -95,6 +95,15 @@ const config: HardhatUserConfig = {
           browserURL: "https://sepolia.arbiscan.io/",
         },
       },
+      {
+        network: "polygonAmoy",
+        chainId: 80002,
+        urls: {
+          apiURL:
+            "https://www.oklink.com/api/explorer/v1/polygonamoy/contract/verify/async",
```
Comment (sanderpick) [inline hardhat.config.ts:103]: i found this on some SO issue... might not be right

File: scripts/utils.ts [hunk 7]
```diff
@@ -1,5 +1,61 @@
...
+  blockNumber: number;
+}
+
+/**
+ * Fetches the current gas fee data from the Amoy Gas Station API.
+ * @param url Amoy gas station API URL.
+ * @returns Current gas fee information or `null` (to align with ethers
+ * `getFeeData` method return type)
+ */
+export async function getFeeData(chainName: string): Promise<FeeData> {
```
Comment (sanderpick) [inline scripts/utils.ts:32]: nice!

File: hardhat.config.ts (hunk 6 above)
Comment (dtbuchholz) [reply hardhat.config.ts:103]: ah, yeah should be `https://www.oklink.com/api/v5/explorer/contract/verify-source-code-plugin/AMOY_TESTNET`. still can't get it to work tho:
```
Verifying on 'polygon-amoy'...
Verifying implementation: 0x3BDC3E46d8de7A2e67A07Ac917382dB6D94577dc
Successfully submitted source code for contract
contracts/TablelandTables.sol:TablelandTables at 0x3BDC3E46d8de7A2e67A07Ac917382dB6D94577dc
for verification on the block explorer. Waiting for verification result...

We tried verifying your contract TablelandTables without including any unrelated one, but it failed.
Trying again with the full solc input used to compile and deploy it.
This means that unrelated contracts may be displayed on Etherscan...

Successfully submitted source code for contract
contracts/TablelandTables.sol:TablelandTables at 0x3BDC3E46d8de7A2e67A07Ac917382dB6D94577dc
for verification on the block explorer. Waiting for verification result...

Failed to verify implementation contract at 0x3BDC3E46d8de7A2e67A07Ac917382dB6D94577dc: The contract verification failed.
Reason: Fail - Unable to verify

Verifying proxy: 0x170fb206132b693e38adFc8727dCfa303546Cec1
Error: Etherscan API call failed with status 500, response: {"code":"50026","msg":"System error, please try again later.","data":[]}
```

Comment (sanderpick) [reply hardhat.config.ts]: hmm, 500 error should mean it's an error on oklink's side, but who knows. is the `browserUrl` correct? 

did you by change compile some changes after deploying? https://github.com/NomicFoundation/hardhat/issues/1349#issuecomment-983002815

Comment (dtbuchholz) [reply hardhat.config.ts]: hm, i didn't change the hardhat compiler version or any of the build settings, so i don't think it's that. but, if i change the compiler version to 0.8.20 and try to verify, it confirms that 0.8.19 _is_ the correct version for the contract i deployed:
> The contract you want to verify was compiled with solidity 0.8.19, but your configured compiler version is: 0.8.20.

based on the polygon discord, others are having the same issue with oklink. however, the polygon team noted [polygonscan will be supporting amoy](https://discord.com/channels/893151281848406016/1114418092601249943/1229463935275630735) within the next 2-3 weeks, so it's probably a moot point.

Comment (sanderpick) [reply hardhat.config.ts]: cool. sg!

**Review by sanderpick — APPROVED** [review]

File: hardhat.config.ts (hunk 6 above)
Comment (iPaulPro) [reply hardhat.config.ts]: https://amoy.polygonscan.com/ is up.

Comment (dtbuchholz) [reply hardhat.config.ts]: @iPaulPro ahh nice, thanks. much better, will add

**Pull request merged**